eth/api.go

```diff
@@ -1771,56 +1774,65 @@ func formatError(err error) string {
-	tx, _, blockIndex, _ := core.GetTransaction(s.eth.ChainDb(), txHash)
-
+func (s *PrivateDebugAPI) TraceTransaction(txHash common.Hash, logger *vm.LogConfig) (*ExecutionResult, error) {
+	if logger == nil {
+		logger = new(vm.LogConfig)
+	}
+	// Retrieve the tx from the chain and the containing block
+	tx, blockHash, _, txIndex := core.GetTransaction(s.eth.ChainDb(), txHash)
 	if tx == nil {
-		return nil, fmt.Errorf("Transaction not found")
+		return nil, fmt.Errorf("transaction %x not found", txHash)
 	}
-
-	block := s.eth.BlockChain().GetBlockByNumber(blockIndex - 1)
+	block := s.eth.BlockChain().GetBlock(blockHash)
 	if block == nil {
-		return nil, fmt.Errorf("Unable to retrieve prior block")
+		return nil, fmt.Errorf("block %x not found", blockHash)
 	}
-
-	// Create the state database
-	stateDb, err := state.New(block.Root(), s.eth.ChainDb())
-	if err != nil {
-		return nil, err
+	// Create the state database to mutate and eventually trace
+	parent := s.eth.BlockChain().GetBlock(block.ParentHash())
+	if parent == nil {
+		return nil, fmt.Errorf("block parent %x not found", block.ParentHash())
 	}
-
-	txFrom, err := tx.FromFrontier()
-
+	stateDb, err := state.New(parent.Root(), s.eth.ChainDb())
 	if err != nil {
-		return nil, fmt.Errorf("Unable to create transaction sender")
-	}
-	from := stateDb.GetOrNewStateObject(txFrom)
-	msg := callmsg{
-		from:     from,
-		to:       tx.To(),
-		gas:      tx.Gas(),
-		gasPrice: tx.GasPrice(),
-		value:    tx.Value(),
-		data:     tx.Data(),
+		return nil, err
 	}
-
-	vmenv := core.NewEnv(stateDb, s.config, s.eth.BlockChain(), msg, block.Header(), vm.Config{
-		Debug:  true,
-		Logger: logger,
-	})
-	gp := new(core.GasPool).AddGas(block.GasLimit())
-
-	ret, gas, err := core.ApplyMessage(vmenv, msg, gp)
-	if err != nil {
-		return nil, fmt.Errorf("Error executing transaction %v", err)
+	// Mutate the state and trace the selected transaction
+	for idx, tx := range block.Transactions() {
+		// Assemble the transaction call message
+		from, err := tx.FromFrontier()
+		if err != nil {
+			return nil, fmt.Errorf("sender retrieval failed: %v", err)
+		}
+		msg := callmsg{
+			from:     stateDb.GetOrNewStateObject(from),
+			to:       tx.To(),
+			gas:      tx.Gas(),
+			gasPrice: tx.GasPrice(),
+			value:    tx.Value(),
+			data:     tx.Data(),
+		}
+		// Mutate the state if we haven't reached the tracing transaction yet
+		if uint64(idx) < txIndex {
+			vmenv := core.NewEnv(stateDb, s.config, s.eth.BlockChain(), msg, parent.Header(), vm.Config{})
+			_, _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(tx.Gas()))
+			if err != nil {
+				return nil, fmt.Errorf("mutation failed: %v", err)
+			}
+			continue
+		}
+		// Otherwise trace the transaction and return
+		vmenv := core.NewEnv(stateDb, s.config, s.eth.BlockChain(), msg, parent.Header(), vm.Config{Debug: true, Logger: *logger})
+		ret, gas, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(tx.Gas()))
```

Comment:
Well, I could either create a gas pool for the entire block and reuse it for all the transactions. Or I can set the available gas to the mount requested by a single transaction and reset after each invocation. I chose this latter one as there's less room for things to go wrong as it's more restrictive, however given that the transaction is already in our database, it should already be valid, so it shouldn't be possible to exceed the tx gas limit.